ws.c

```diff
@@ -0,0 +1,409 @@
+		evws_connection_free(ctx);
+	}
+}
+
+static void
+close_event_cb(struct bufferevent *bev, short what, void *ctx)
+{
+	evws_connection_free(ctx);
+}
+
+void
+evws_close(struct evws_connection *evws, uint16_t reason)
+{
+	uint8_t fr[4] = {0x8 | 0x80, 2, 0};
+	struct evbuffer *output;
+	uint16_t *u16;
+
+	if (evws->closed)
+		return;
+	evws->closed = true;
+
+	u16 = (uint16_t *)&fr[2];
+	*u16 = htons((int16_t)reason);
+	output = bufferevent_get_output(evws->bufev);
+	evbuffer_add(output, fr, 4);
+
+	/* wait for close frame writing complete and close connection */
+	bufferevent_setcb(
+		evws->bufev, NULL, close_after_write_cb, close_event_cb, evws);
+}
+
+static void
+evws_force_disconnect_(struct evws_connection *evws)
+{
+	evws_close(evws, WS_CR_NONE);
+}
+
+static enum WebSocketFrameType
+get_ws_frame(unsigned char *in_buffer, int buf_len, unsigned char **payload_ptr,
+	int *out_len)
+{
+	unsigned char opcode;
+	unsigned char fin;
+	unsigned char masked;
+	int payload_len;
+	int pos;
+	int length_field;
+	unsigned int mask;
+
+	if (buf_len < 3) {
+		return INCOMPLETE_FRAME;
+	}
+
+	opcode = in_buffer[0] & 0x0F;
+	fin = (in_buffer[0] >> 7) & 0x01;
+	masked = (in_buffer[1] >> 7) & 0x01;
+
+	payload_len = 0;
+	pos = 2;
+	length_field = in_buffer[1] & (~0x80);
+	mask = 0;
+
+	if (length_field <= 125) {
+		payload_len = length_field;
+	} else if (length_field == 126) { /* msglen is 16bit */
+		payload_len = ntohs(*(uint16_t *)(in_buffer + 2));
+		pos += 2;
+	} else if (length_field == 127) { /* msglen is 64bit */
+		payload_len = ntohs(*(uint64_t *)(in_buffer + 2));
+		pos += 8;
+	}
+	if (buf_len < payload_len + pos) {
+		return INCOMPLETE_FRAME;
+	}
+
+	if (masked) {
+		unsigned char *c;
+
+		mask = *((unsigned int *)(in_buffer + pos));
+		pos += 4;
+
+		/* unmask data */
+		c = in_buffer + pos;
+		for (int i = 0; i < payload_len; i++) {
+			c[i] = c[i] ^ ((unsigned char *)(&mask))[i % 4];
+		}
+	}
+
+	*payload_ptr = in_buffer + pos;
+	*out_len = payload_len;
+
+	switch (opcode) {
+	case 0x0:
+		return fin ? TEXT_FRAME : INCOMPLETE_TEXT_FRAME;
+	case 0x1:
+		return fin ? TEXT_FRAME : INCOMPLETE_TEXT_FRAME;
+	case 0x2:
+		return fin ? BINARY_FRAME : INCOMPLETE_BINARY_FRAME;
+	case 0x9:
+		return PING_FRAME;
+	case 0xA:
+		return PONG_FRAME;
+	default:
+		return ERROR_FRAME;
+	}
+}
+
+
+static void
+ws_evhttp_read_cb(struct bufferevent *bufev, void *arg)
+{
+	struct evws_connection *evws = arg;
+	unsigned char *payload;
+	enum WebSocketFrameType type;
+	int msg_len, in_len;
+	struct evbuffer *input = bufferevent_get_input(evws->bufev);
+
+	while ((in_len = evbuffer_get_length(input))) {
+		unsigned char *data = evbuffer_pullup(input, in_len);
+
+		type = get_ws_frame(data, in_len, &payload, &msg_len);
+
+		if (type == TEXT_FRAME) {
+			if (evws->cb) {
+				char *text = mm_malloc(msg_len + 1);
+
+				memcpy(text, payload, msg_len);
+				text[msg_len] = 0;
+				evws->cb(evws, text, msg_len, evws->cb_arg);
```

Comment:
Could you give me an example how to properly use it in this specific case? Previously it was Linux-only implementation and I used a lot of `alloca()` but it seems is not portable

Reply:
>Could you give me an example how to properly use it in this specific case?

Actually I don't think that `evbuffer` is good idea here. How about passing `const char *` (to add a sign to user that the data is const/immutable), the reference to existing array, instead of allocating new?

TL;DR;

I thought about some variants, and I cannot find better way than simply pass `const char *`.

Creating temporary `evbuffer` to pass to the callback:
- too costly, 144 (evbuffer) + 48 (evbuffer_chain) is big overhead
- and even if you will ignore overhead, it still does not have any benefits over `const char *`, since if it will hold the reference, then anyway if user want to, he could change it. By the same reason per-connection `evbuffer` will not help.

But just to show you idea, here is a hacky patch -
[ws-evbuffer.patch.txt](https://github.com/libevent/libevent/files/9455518/ws-evbuffer.patch.txt)

> Previously it was Linux-only implementation and I used a lot of alloca() but it seems is not portable

`alloca` is indeed not good to use here.

Reply:
Now I got it, my previous idea was to provide a simple API for library user to get WS text message with null terminator and then user just can use it as a const char* in its code
Obviously, there is no way to do this without additional allocations, but for me using `struct evbuffer *` in message callback looks overcomplicated (because of necessity of using `evbuffer_pullup` to extract actual data pointer where we can use just raw pointer and message data length - there is no way of delivering incomplete WS message now)

Reply:
Just tested [ws-evbuffer.patch.txt](https://github.com/libevent/libevent/files/9455518/ws-evbuffer.patch.txt)`, this not will work because actual input buffer can contain multiple frames at the same time (that why we're using `while` loop)

Reply:
Please check latest implementation, I used your evbuffer ideas (first it looked complicated, but I found code had no proper support for handling incomplete frames), implemented as well with binary frames support

Reply:
That what we need, though a little bit unsafe, but I guess it is fine!